### PR TITLE
Add internal/luksview package

### PR DESCRIPTION
**Pull request by chrisccoulson**
This adds a new package containing a view of the LUKS2 metadata
in a way that is most useful to secboot.

It also contains a representation of tokens that will be used by
secboot. Each token has a name, which is an arbitrary string used to
identify its associated keyslot. The name is intended to be unique
between keyslots and secboot will enforce this. All tokens created by
secboot will have a name and will be associated with one keyslot.

Tokens without a name or which are associated by more than one keyslot
are ignored by the View API.

For context, here's some notes about how the View API is intended to be
used:
- View.TokenByName returns the token for the keyslot with the
supplied name, and is intended to be used when performing changes to
the slot (eg, updating the PCR policy in the TPM case or deleting the
slot).
- View.KeyDataTokensByPriority returns a list of key data tokens
in priority order, and is intended to be used internally by the
ActivateVolumeWithKeyData API. There is no priority for recovery slots
as it doesn't make sense - attempting each slot in a defined order
rather than letting cryptsetup pick the order would require repeated
execution of the KDF if more than one recovery keyslot is defined.
- View.OrphanedTokenIds returns a list of IDs associated with
tokens that have been orphaned (ie, their associated keyslot has been
deleted), and can occur if the process of deleting a keyslot and its
associated token is interrupted. This can be used to resume an
interrupted removal.
- View.UsedKeyslots returns a list of the currently used keyslot IDs.
This will be used by secboot to manually pick an ID for new keyslots
rather than letting cryptsetup handle it, as this makes it easier to
subsequently create a token associated with it.

### CODE REVIEW

File: internal/luksview/view.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	}
+	if _, orphaned := data.token.(*orphanedToken); orphaned {
+		return nil, 0, false
+	}
+	return data.token, data.id, true
+}
+
+// KeyDataTokensByPriority returns all of the key data tokens in order of priority,
+// from highest to lowest. Tokens with the same priority are returned in the order in
+// which their names are sorted. This omits any with a priority of 0.
```
Comment (pedronis) [inline internal/luksview/view.go:138]: the code and the comment don't seem to match. The code has priority < 0 (not <= 0)

**Review by pedronis — COMMENTED** [review]
overall looks good, some naming and detail remarks

File: internal/luksview/tokens.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+	return nil
+}
+
+func (t *orphanedToken) Name() string {
+	return t.raw.Name
+}
+
+// NewOrphanedTokenForTesting returns a new orphaned named token with the
+// supplied type and name, which is useful for testing.
+func NewOrphanedTokenForTesting(t luks2.TokenType, name string) NamedToken {
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply internal/luksview/tokens.go]: I've renamed this now

File: internal/luksview/view.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	view := &View{source: source}
+
+	if err := view.Refresh(); err != nil {
+		return nil, err
+	}
+
+	return view, nil
+}
+
+// Refresh updates this view from the source container.
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply internal/luksview/view.go]: Reread sounds ok

File: internal/luksview/view.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+		}
+
+		v.namedTokens[named.Name()] = namedTokenData{id: id, token: named}
+	}
+
+	return nil
+}
+
+// ListNames returns a sorted list of all of the keyslot names from this view.
+func (v *View) ListNames() (names []string) {
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply internal/luksview/view.go]: I've renamed this to TokenNames

File: internal/luksview/view.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+			continue
+		}
+		names = append(names, name)
+	}
+	sort.Strings(names)
+	return names
+}
+
+// TokenByName returns the token and its ID for the keyslot with the supplied name.
+func (v *View) TokenByName(name string) (token NamedToken, id int, exists bool) {
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply internal/luksview/view.go]: I've changed this to inUse. I noticed also that an orphaned token would occupy a name but wouldn't be returned from TokenByName or TokenNames, so I've corrected that as well.

File: internal/luksview/view.go (hunk 1 above)
Comment (chrisccoulson) [reply internal/luksview/view.go]: I've updated this.

File: internal/luksview/tokens.go [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+
+type recoveryTokenRaw struct {
+	tokenBaseRaw
+}
+
+type keyDataTokenRaw struct {
+	tokenBaseRaw
+	Priority int             `json:"ubuntu_fde_priority"`
+	Data     json.RawMessage `json:"ubuntu_fde_data,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply internal/luksview/tokens.go]: I've moved these now.

**Pull request merged**